go/database/mpt/proof.go

```diff
@@ -0,0 +1,116 @@
+package mpt
+
+import (
+	"github.com/Fantom-foundation/Carmen/go/common"
+)
+
+//go:generate mockgen -source proof.go -destination proof_mocks.go -package witness
+
+// RlpEncodedNode is an RLP encoded MPT node.
+type RlpEncodedNode []byte
+
+// proofDb is a database of RLP encoded MPT nodes and their hashes that represent witness proofs.
+type proofDb map[common.Hash]RlpEncodedNode
+
+// WitnessProof represents a witness proof.
+// It contains a database of MPT nodes and their hashes.
+// The proof may be verified by iterating the proof database from the
+// root hash to a terminal entry according to a Path
+// representing either an account address or a storage key.
+// The proof is valid if it can be iterated to an expected terminal entry.
+type WitnessProof proofDb
+
+// Merge merges the input witness proof into the current witness proof.
+func (p *WitnessProof) Merge(other *WitnessProof) {
+	for k, v := range *other {
+		(*p)[k] = v
+	}
+}
+
+// ExtractAccountProof extracts a sub-proof from the current witness proof.
+// It returns a copy that contains only proof for the input address and storage keys.
+// If the input address or keys are not found in the current proof, it returns an empty proof.
```

Comment:
If some keys are not provable, will the resulting proof cover those or will it be anyway empty?

Reply:
I will update the documentation. I think it is not correct at the moment. I think it should work similar to proving account/keys, it will keep adding nodes into the proof as far as it can reach. If the account or a key does not lead to a result, the proof will contain only nodes that could be reach. It can be empty if no node was reached.

Reply:
but not sure if such a partial proof is useful?

Reply:
I think such a proof is useful, since it proofs that the key's value is 0.

Reply:
similar question to the one below: should we return bool to inform that the input was not proven? Or do we just return the proof, and the user is expected to call one of the getters to actually see that the input is not proven?

Reply:
The only way it can not be proven is if there was an IO error. In this case an error is already expected. No need for a bool.